mvg/utils/response_processing.py

```diff
@@ -40,25 +42,28 @@ def get_paginated_items(request: Callable, url: str, params: Dict) -> Dict:
         - "items": list of int, representing timestamps.
         - "total": int, representing total number of items.
     """
-    if "limit" in params or "offset" in params:
-        response = request("get", url, params=params)
-        response = response.json()
-        all_items = response["items"]
-        num_items = response["total"]
-    else:
-        # List all by default if pagination is not requested
-        response = request("get", url, params=params)
-        resp_first = response.json()
-        all_items = resp_first["items"]
-        num_items = resp_first["total"]
-        limit = resp_first["limit"]
-        num_reqs = (num_items - 1) // limit
-        for idx in range(1, num_reqs + 1):
-            offset = idx * limit
+    response = request("get", url, params=params)
+    response = response.json()
+    items = response["items"]
+    total_items_num = response["total"]
+    request_limit = response["limit"]
+
+    offset = params.get("offset", 0)
+    limit = params.get("limit", total_items_num)
+    expected_num = clamp(limit, 0, total_items_num - offset)
+    missing_num = expected_num - len(items)
+
+    if missing_num > 0:
+        num_reqs = (missing_num - 1) // request_limit + 1
+        for _ in range(0, num_reqs):
+            if "limit" in params:
+                params["limit"] -= request_limit
+            offset += request_limit
             params["offset"] = offset
             response = request("get", url, params=params)
-            all_items += response.json()["items"]
-    return {"items": all_items, "total": num_items}
+            items += response.json()["items"]
+
+    return {"items": items, "total": total_items_num}
```

Comment:
This has become complex now. There should be a better way to write this up.

Reply:
I could not reasonably simplify the solution to satisfy current requirements. If you have a solution in mind which you consider a better option could you submit it as a follow-up refactoring after this PR? Or if you think you can do this relatively quickly just as a commit to this PR.